Sources/Brave/Frontend/Browser/Search/Voice Search/SpeechRecognizer.swift

```diff
@@ -0,0 +1,300 @@
+      return
+    }
+    
+    do {
+      let (audioEngine, request) = try setupStartEngine()
+        
+      self.audioEngine = audioEngine
+      self.request = request
+      
+      task = recognizer.recognitionTask(with: request, resultHandler: { [weak self] result, error in
+        guard let self = self else {
+          return
+        }
+        
+        var isFinal = false
+        
+        if let result, !isFinal {
+          let formattedTranscript = result.bestTranscription.formattedString
+          let transcriptComponents = formattedTranscript.components(separatedBy: .whitespacesAndNewlines)
+          let formattedWords = transcriptComponents.filter { !$0.isEmpty }
+          
+          // SpeechRecognitionMetadata is the key to detect speaking finalized
+          isFinal = result.isFinal
+          || result.speechRecognitionMetadata != nil
+          || formattedWords.count >= 15
+              
+          if !formattedTranscript.isEmpty {
+            self.transcribe(formattedTranscript)
+          }
+        }
+        
+        // Check voice input final
+        if isFinal {
+          animationType = .stable
+          transcriptedIcon = "leo.check.circle-outline"
+          
+          // Remove audio buffer input
+          audioEngine.inputNode.removeTap(onBus: 0)
+          // Reset Speech Recognizer
+          self.reset()
+
+          Task.delayed(bySeconds: 0.75) { @MainActor in
+            if let formattedTranscript = result?.bestTranscription.formattedString, !formattedTranscript.isEmpty {
+              self.finalize(searchQuery: formattedTranscript)
+            }
+          }
+        }
+      })
+    } catch {
+      reset()
+      log.debug("Voice Search Recognization Fault \(error.localizedDescription)")
+    }
+  }
+    
+  /// Reset the speech recognizer.
+  private func reset() {
+    try? AVAudioSession.sharedInstance().setActive(false)
+    task?.cancel()
+    audioEngine?.stop()
+    
+    audioEngine = nil
+    request = nil
+    task = nil
+  }
+    
+  private func setupStartEngine() throws -> (AVAudioEngine, SFSpeechAudioBufferRecognitionRequest) {
+    let audioEngine = AVAudioEngine()
+    
+    let request = SFSpeechAudioBufferRecognitionRequest()
+    request.shouldReportPartialResults = true
+    request.requiresOnDeviceRecognition = true
+
+    let audioSession = AVAudioSession.sharedInstance()
+    try audioSession.setCategory(.playAndRecord, mode: .measurement, options: .duckOthers)
+    try audioSession.setActive(true, options: .notifyOthersOnDeactivation)
+    let inputNode = audioEngine.inputNode
+    
+    let recordingFormat = inputNode.outputFormat(forBus: 0)
+    inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { [weak self] buffer, when in
+      guard let self else { return }
+      
+      request.append(buffer)
+      
+      guard let channelData = buffer.floatChannelData?[0] else {
```

Comment:
This 0 index is safe, wont crash?

Reply:
Yes it is is entirely safe. 

From floatChannelData documentation:

The returned pointer is to format.channelCount pointers to float. Each of these pointers is to samples.

Format is deinterleaved,  the pointers point to separate chunks of memory, and  `property value is 1`.
Format is in an interleaved state, the pointers refer to the same buffer of interleaved samples, each offset by 1 frame, and the property value is the `number of interleaved channels`.